source/data-handling-bigfiles-R.Rmd

```diff
@@ -0,0 +1,240 @@
+minDate = strptime('1/3/2014',format = '%d/%m/%Y')
+maxDate = strptime('1/10/2014',format = '%d/%m/%Y')
+
+streamFile <- function(limit) {
+    con <- file(csv.name, open = "r")
+    selectedRecords <- list()
+    i <- 0
+    file.streaming.timing <- system.time(
+        while (i < limit) {
+            oneLine <- readLines(con, n = 1, warn = FALSE)
+            vec = (strsplit(oneLine, ","))
+            selectedRecords <- c(selectedRecords, vec)
+            i <- i + 1
+        }
+    )
+    close(con)
+    return(file.streaming.timing[[3]])
+}
+
+freadFile <- function(limit) {
+    file.fread.timing = system.time(
+        d <- fread(csv.name, showProgress = FALSE, nrows = limit)
+    )
+    return(file.fread.timing[[3]])
+}
+
+maxLines <- c(5000, 10000, 15000, 20000, 25000, 30000)
+streamingTimes <- sapply(maxLines, streamFile)
+freadTimes <- sapply(maxLines, freadFile)
+data <- data.frame(n = maxLines, streaming = streamingTimes, 
+                   fread = freadTimes)
+pdata <- melt(data, id = c("n"))
+colnames(pdata) <- c("n", "algorithm", "execTime")
+qplot(n, execTime, data = pdata, color = algorithm, 
+      xlab = "number of lines read", ylab = "execution time (s)")
+```
+
+# The database file strategy
+
+## Working with SQLite databases
+
+SQLite databases are single file databases meaning you can simply download them, store them in a folder or share them with colleages. Similar to a csv. They are however more powerful than csv's because of two important features:
+
+* Support for SQL: this allows you to execute intelligent filters on your data, similar to the `sqldf` package or database environments you are familiar with. That way, you can reduce the amount of data that's stored in memory by filtering out rows or columns.
+* Indexes: SQLite databases contain indexes. An index is something like an ordered version of a column. When enabled on a column, you can search through the column very fast. We'll demonstrate this below.
+
+Here we've downloaded a second file `2016-04-20-processed-logs-big-file-example.db` that contains the same data as the `2016-04-20-processed-logs-big-file-example.csv` file, but as a sqlite database. Furthermore, the database contains indexes which will dramatically drop the time needed to perform search queries. If you don't have a SQLite database containing your data, try creating one as described [here](https://cran.r-project.org/web/packages/dplyr/vignettes/databases.html).
+
+Let's first connect to the database and list the available tables.
+
+```{r sqliteconnect}
+library(RSQLite)
+db <- dbConnect(SQLite(), dbname = db.name)
+
+# show the tables in this database
+dbListTables(db)
+```
+
+Let's try to select rows where the device id matches a given value (e.g. 860), and the date time is between two given timestamps. For our analysis, we only need date_time, latitude, longitude and altitude so we only select those.
+
+```{r}
+sqlTiming <- system.time(data <-  dbGetQuery(conn = db,
```

Comment:
Thanks for the advice. Probably out of scope for this comparison, as the differences are clear enough.